File: Go/euler-216-the-primality.go/euler-216-the-primality.go
package main

// Importing necessary packages
import (
	"fmt"     // Provides formatted I/O functions
	"os"      // Provides a way to interact with the operating system, including command-line arguments
	"strconv" // Provides functions for string to integer conversion
)

var totalCount = 0

// primality function checks the primality of certain terms in a sequence
// and updates the totalCount of prime terms found.
func primality(limit int) int {
	// Generate the sequence of terms
	sequence := generateSequence(limit)

	// Iterate through the sequence and check for primality
	for i, term := range sequence {
		// Skip the first two terms as they are set to -1
		if i < 2 {
			continue
		}

		// Check if the term is prime using a specific formula
		if term == int64(2*i*i-1) {
			totalCount++
		}

		// If the term is within the limit, perform factorization
		if term <= int64(limit*2) && 1 < term {
			currentTerm := term

			// Divide out factors in the positive direction
			divideOutFactors(sequence, currentTerm, i, limit)
			// Divide out factors in the negative direction
			divideOutFactorsNegative(sequence, currentTerm, i, limit)
		}
	}

	return totalCount
}

// divideOutFactors performs factorization in the positive direction
func divideOutFactors(sequence []int64, currentTerm int64, i, limit int) {
	term := int(currentTerm)

	// Iterate through the sequence and divide out factors
	for j := i + term; j <= limit; j += term {
		// Continue dividing out factors until the currentTerm is no longer a factor
		for sequence[j]%currentTerm == 0 {
			sequence[j] /= currentTerm
		}
	}
}

// divideOutFactorsNegative performs factorization in the negative direction
func divideOutFactorsNegative(sequence []int64, currentTerm int64, i, limit int) {
	term := int(currentTerm)

	// Iterate through the sequence and divide out factors in the negative direction
	for j := i + (term-i)*2%term; j <= limit; j += term {
		// Continue dividing out factors until the currentTerm is no longer a factor
		for sequence[j]%currentTerm == 0 {
			sequence[j] /= currentTerm
		}
	}
}

// generateSequence generates a sequence of terms based on a given limit
func generateSequence(limit int) []int64 {
	sequence := make([]int64, limit+1)

	// Populate the sequence using a specific formula
	for i := 2; i <= limit; i++ {
		sequence[i] = int64(2*i*i - 1)
	}

	// Set initial terms to -1
	sequence[0], sequence[1] = -1, -1

	return sequence
}

func main() {
	// Check if the correct number of command-line arguments is provided
	if len(os.Args) != 2 {
		fmt.Println("Usage: go run euler-216-the-primality.go <limit>")
		return
	}

	// Parse the limit from the command-line argument
	limit, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Printf("Invalid input: Please enter a valid limit.\nError message: %v\n", err)
		return
	}

	// Call the primality function and print the result
	final := primality(limit)
	fmt.Println(final)
}

File: C++/euler-216-the-primality.cpp/euler-216-the-primality.cpp
#include <iostream>   // Input and output stream handling (for std::cout and std::endl)
#include <vector>     // Dynamic array for storing prime factors in the HammingNumbers class
#include <cstdint>    // Standard integer types for fixed-size integers, used for int64_t as integers have larger values in hammingNumbers

// Class to handle primality calculations
class Primality {
private:
    std::vector<long long> sequence;  // Sequence of terms
    int totalCount;                   // Total count of prime terms found

    // Function to perform factorization in the positive direction
    void divideOutFactorsPositive(int currentTerm, int i, int limit) {
        for (int j = i + currentTerm; j <= limit; j += currentTerm) {
            // Continue dividing out factors until the currentTerm is no longer a factor
            while (sequence[j] % currentTerm == 0) {
                sequence[j] /= currentTerm;
            }
        }
    }

    // Function to perform factorization in the negative direction
    void divideOutFactorsNegative(int currentTerm, int i, int limit) {
        for (int j = i + (currentTerm - i) * 2 % currentTerm; j <= limit; j += currentTerm) {
            // Continue dividing out factors until the currentTerm is no longer a factor
            while (sequence[j] % currentTerm == 0) {
                sequence[j] /= currentTerm;
            }
        }
    }

    // Function to generate a sequence of terms based on a given limit
    void genSequence(int limit) {
        sequence.resize(limit + 1);

        // Populate the sequence using a specific formula
        for (int i = 2; i <= limit; ++i) {
            sequence[i] = 2LL * i * i - 1;
        }

        // Set initial terms to -1
        sequence[0] = sequence[1] = -1;
    }

public:
    // Constructor to initialize totalCount
    Primality() : totalCount(0) {}

    // Function to check primality of terms and update totalCount
    int primality(int limit) {
        // Generate the sequence of terms
        genSequence(limit);

        // Iterate through the sequence and check for primality
        for (int i = 2; i <= limit; ++i) {
            // Check if the term is prime using a specific formula
            if (sequence[i] == 2LL * i * i - 1) {
                totalCount++;
            }

            // If the term is within the limit, perform factorization
            if (sequence[i] <= limit * 2 && sequence[i] > 1) {
                int currentTerm = sequence[i];

                // Divide out factors in the positive direction
                divideOutFactorsPositive(currentTerm, i, limit);
                // Divide out factors in the negative direction
                divideOutFactorsNegative(currentTerm, i, limit);
            }
        }

        return totalCount;
    }
};

int main(int argc, char* argv[]) {
    // Check if the correct number of command-line arguments is provided
    if (argc != 2) {
        std::cout << "Usage: ./euler-216-the-primality <limit>" << std::endl;
        return 0;
    }

    char* endptr;


    // Convert command line arguments to integers
    int limit = std::strtol(argv[1], &endptr, 10);

    // Check if the conversion was successful
    if (*endptr != '\0') {
        std::cout << "Error: <limit> must be integers." << std::endl;
        return 0;
    }

    // Create an instance of the Primality class
    Primality primality;

    // Call the primality function and print the result
    int final = primality.primality(limit);
    std::cout << final << std::endl;

    return 0;
}
